### PR TITLE
MNT: use pythoncapi_compat.h in npy_compat.h

### PR DESCRIPTION
**Pull request by ngoldbaum**
Following on from #26158 and #26159, this adds a new submodule for `pythoncapi_compat.h` and includes it in `npy_pycompat.h`. I also removed all the spots where `npy_pycompat.h` was getting included unnecessarily, probably leftovers from the python 2->3 migrations.

This doesn't actually use the new header yet, that will come in followups.

### CODE REVIEW

**Comment by seberg** [comment]
In the past we used very few macro definitions and that also felt much to use a submodule (and the full file which has a lot of things we don't need).

I suspect we still only use a small, but significantly larger, fraction of them all and we have a lot of submodules now anyway.
So I lean towards this making sense even if we end up using very few functions, there are probably some other small nice things which we get this way without having to think about it (i.e. `NewRef`).

**Review by mhvk — APPROVED** [review]
Yes, seems a sensible way to go. I guess eventually the submodule will disappear again, once borrowed references are a thing of the past? And that any actual changes in the code to use new routines that give new references will be done in follow-up? 

p.s. I had to do a bit of a double-take since you seemed to remove the usage of the headers that you were introducing, but that just meant I should have read your comments better...

**Comment by ngoldbaum** [comment]
> I guess eventually the submodule will disappear again, once borrowed references are a thing of the past?

Maybe but I don’t see the CPython C API remaining static after the nogil effort over the next few years, and this way we don’t need to maintain our own compatibility shims. I also like these shims because their names match the upstream names so they don’t need another migration in the future to get rid of them besides removing the compat header include.

**Review by rgommers — COMMENTED** [review]
This LGTM overall, thanks @ngoldbaum. `_core/src/common` seems like the right place to put this. I have one suggestion to avoid contributors being puzzled after a `git pull` and missing the submodule - add this near the top of `numpy/_core/meson.build`:
```meson
  if not fs.exists('src/common/pythoncapi-compat')
    error('Missing the `pythoncapi-compat` git submodule! Run `git submodule update --init` to fix this.')
  endif
```
There are a couple of other checks like that already, like for `svml` and `highway`.

File: numpy/_core/src/umath/override.c [hunk 1]
```diff
@@ -1,7 +1,7 @@
 #define NPY_NO_DEPRECATED_API NPY_API_VERSION
-#define NO_IMPORT_ARRAY
+#define NO_IMPORT_
```
Comment (seberg) [inline numpy/_core/src/umath/override.c:2]: This looks like an accidental change.

Comment (ngoldbaum) [reply numpy/_core/src/umath/override.c:2]: Good eye, thanks!

**Review by rgommers — APPROVED** [review]
Reviewers seem happy and CI seems happy, so let's give it a go:) Thanks Nathan, all!

**Pull request merged**